casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -74,6 +86,171 @@ class RuntimeManagerImpl[F[_]: Concurrent: Metrics: Span: Log](
+    import coop.rchain.rspace.util._
+    (if (replay) runtime.replaySpace else runtime.space)
+      .consume(
+        Seq(systemDeploy.returnChannel),
+        Seq(systemDeployConsumeAllPattern),
+        emptyContinuation,
+        persist = false
+      )
+      .map(unpackOption)
+  }
+
+  def replaySystemDeploy[S <: SystemDeploy](stateHash: StateHash)(
+      systemDeploy: S,
+      processedSystemDeploy: ProcessedSystemDeploy
+  ): F[Either[ReplayFailure, SystemDeployReplayResult[systemDeploy.Result]]] =
+    withRuntimeLock { runtime =>
+      runtime.replaySpace.reset(Blake2b256Hash.fromByteString(stateHash)) >>
+        replaySystemDeployInternal(runtime)(systemDeploy, processedSystemDeploy) >>= {
+        case Right(Right(result)) =>
+          runtime.replaySpace
+            .createCheckpoint()
+            .map(_.root.toByteString)
+            .map(
+              finalStateHash =>
+                SystemDeployReplayResult.replaySucceeded(finalStateHash, result).asRight
+            )
+        case Right(Left(systemDeployError)) =>
+          SystemDeployReplayResult.replayFailed(systemDeployError).asRight.pure
+        case Left(replayFailure) =>
+          replayFailure.asLeft.pure
+      }
+    }
+
+  private def replaySystemDeployInternal[S <: SystemDeploy](runtime: Runtime[F])(
+      systemDeploy: S,
+      processedSystemDeploy: ProcessedSystemDeploy
+  ): F[Either[ReplayFailure, Either[SystemDeployError, systemDeploy.Result]]] = {
+
+    def verifyReplaySuccess(
+        result: systemDeploy.Result
+    ): Either[ReplayFailure, systemDeploy.Result] =
+      if (processedSystemDeploy.failed)
+        ReplayFailure.replayStatusMismatch(initialFailed = true, replayFailed = false).asLeft
+      else result.asRight
+
+    def verifyReplayFailure(systemDeployError: SystemDeployError): Option[ReplayFailure] =
+      processedSystemDeploy match {
+        case Succeeded(_) =>
+          ReplayFailure.replayStatusMismatch(initialFailed = false, replayFailed = true).some
+        case Failed(_, errorMsg) =>
+          if (errorMsg == systemDeployError.errorMsg) none
+          else ReplayFailure.systemDeployErrorMismatch(errorMsg, systemDeployError.errorMsg).some
+      }
+
+    runtime.replaySpace.rig(processedSystemDeploy.eventList.map(EventConverter.toRspaceEvent)) >> {
+      runtime.replaySpace.createSoftCheckpoint() >>= { softCheckpoint =>
+        evaluateSystemSource(runtime)(systemDeploy, replay = true) >>= {
+          case EvaluateResult(_, interpreterErrors) =>
+            if (interpreterErrors.nonEmpty) UnexpectedSystemErrors(interpreterErrors).raiseError
+            else
+              consumeResult(runtime)(systemDeploy, replay = true) >>= { consumeResultOpt =>
+                runtime.replaySpace.checkReplayData().attempt.map(_.swap.toOption) >>= {
+                  case Some(replayException: ReplayException) =>
+                    ReplayFailure.unusedCOMMEvent(replayException).asLeft.pure
```

Comment:
Do we need `revertToSoftCheckpoint ` here?

Reply:
I think we should add some comments to the code about when and why we revert or not. It's hard to understand from code.

Reply:
@zsluedem https://github.com/rchain/rchain/pull/2778#discussion_r341574398